src/io.rs

```diff
@@ -0,0 +1,20 @@
+// Rust Bitcoin Library
+// Written in 2014 by
+//   Andrew Poelstra <apoelstra@wpsoftware.net>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+//! no_std polyfill 
+
+#[cfg(feature = "bare-io")]
+pub use bare_io::*;
+
+#[cfg(feature = "std")]
+pub use std::io::*;
```

Comment:
Any ideas on a better way to polyfill `std::io`?

Reply:
What things need to be polyfilled exactly?

I mean at some level we can just feature-gate all stuff that needs std things, but I suppose your concern is that the fraction that would remain is so small that a polyfill could make more things no-std-compatible?

How does a polyfill version of `Vec` work?

Reply:
F.e. if on-stack `Vec`'s have a hard length requirement, that may cause various errors in different places as we assume that Vec's have infinite space. (F.e. when writing/encoding into them, we always `unwrap` errors so a length limit would cause these to panic.)

Reply:
The polyfill is only for `std::io`. We're using `alloc::Vec`, which is exactly the same as `std::Vec`. 

But `HashMap` and `HashSet` don't exist in no_std. As mentioned in the top comment in this PR, I swapped them out for `BTreeMap` and `BTreeSet` *globally* because those exist in no_std and I just wanted to get this working. We could add 3rd crates that implement no_std-compatible `HashMap` and `HashSet` (e.g. [hashbrown](https://crates.io/crates/hashbrown/0.1.0)). Or we could re-export `BTreeSet` as `HashSet` in the polyfill when `std` feature is off.

I'm still wrapping my head around the exact stuff from `std::io` we actually use, and whether we could drop this `bare_io` dependency and just implement what we actually use. Not sure how feasible this is.